src/main/java/org/jvnet/hudson/test/RealJenkinsRule.java

```diff
@@ -0,0 +1,348 @@
+                                                " that you have a clean target directory. Try running 'mvn test-compile' " +
+                                                "from the command line (once only), which will copy the required plugins " +
+                                                "into target/test-classes/test-dependencies - then retry your test", x);
+                                    } else {
+                                        throw new IOException(index + " contains bogus line " + line, x);
+                                    }
+                                }
+                                if (f.exists()) {
+                                    FileUtils.copyURLToFile(url, new File(plugins, line + ".jpi"));
+                                } else {
+                                    FileUtils.copyURLToFile(new URL(index, line + ".hpi"), new File(plugins, line + ".jpi"));
+                                }
+                                // TODO add method to disable a plugin (e.g. to test optional dependencies)
+                            }
+                        }
+                    }
+                    base.evaluate();
+                } finally {
+                    try {
+                        tmp.dispose();
+                    } catch (Exception x) {
+                        LOGGER.log(Level.WARNING, null, x);
+                    }
+                }
+            }
+        };
+    }
+
+    /**
+     * One step to run.
+     * <p>Since this thunk will be sent to a different JVM, it must be serializable.
+     * The test class will certainly not be serializable, so you cannot use an anonymous inner class.
+     * If your thunk requires no parameters from the test JVM, the friendliest idiom is a static method reference:
+     * <pre>
+     * &#64;Test public void stuff() throws Throwable {
+     *     rr.then(YourTest::_stuff);
+     * }
+     * private static void _stuff(JenkinsRule r) throws Throwable {
+     *     // as needed
+     * }
+     * </pre>
+     */
+    @FunctionalInterface
+    public interface Step extends Serializable {
+        void run(JenkinsRule r) throws Throwable;
+    }
+
+    // TODO add methods for more advanced use cases (such as multiple Jenkins services interacting):
+    // · withSession: run a Jenkins session, run a test thunk _locally_ (blackbox testing), shut down
+    // · runRemotely: run a test thunk remotely, given a running session (whitebox testing)
+    // Thus rr.then(s) would be shorthand for rr.withSession(() -> rr.runRemotely(s))
+
+    /**
+     * Run one Jenkins session, send a test thunk, and shut down.
+     */
+    public void then(Step s) throws Throwable {
+        Body.writeSer(new File(home, "step.ser"), s);
+        String cp = System.getProperty("java.class.path");
+        ProcessBuilder pb = new ProcessBuilder(
+                new File(System.getProperty("java.home"), "bin/java").getAbsolutePath(),
```

Comment:
I wonder if this should pickup the JVM memory args from surefire so that it doesn't just allocate random amount of memory?

Or should we make it an option in the actual Rule?

Reply:
https://github.com/jenkinsci/jenkins-test-harness/blob/9f4ca88414b78e8420e0bc55b71b8c4e3ca767b8/src/main/java/org/jvnet/hudson/test/RealJenkinsRule.java#L236 Plenty of room for improvement, consider this an MVP.